private/buf/bufwire/proto_encoding_writer.go

```diff
@@ -60,14 +59,26 @@ func (i *protoEncodingWriter) PutMessage(
 	if err != nil {
 		return err
 	}
-	data, err := protoencoding.NewJSONMarshalerIndent(resolver).Marshal(message)
-	if err != nil {
-		return err
+	var marshaler protoencoding.Marshaler
+	switch messageRef.MessageEncoding() {
+	case bufconvert.MessageEncodingBin:
+		marshaler = protoencoding.NewWireMarshaler()
+	case bufconvert.MessageEncodingJSON:
+		marshaler = protoencoding.NewJSONMarshalerIndent(resolver)
+	default:
+		return fmt.Errorf("unknown message encoding type")
 	}
-	writeCloser, err := i.fetchWriter.PutSingleFile(ctx, container, path)
+	data, err := marshaler.Marshal(message)
 	if err != nil {
 		return err
 	}
+	writeCloser := ioextended.NopWriteCloser(container.Stdout())
+	if messageRef.Path() != "-" {
+		writeCloser, err = os.Create(messageRef.Path())
```

Comment:
This caught my eye at first, but it looks like we do the same thing in our internal `buffetch` writer.

Reply:
yes the underlying logic in `singleRef` is something similar to this